lib/ghtorrent/etag_helper.rb

```diff
@@ -0,0 +1,96 @@
+class EtagHelper
+  def initialize(retriever, url)
+    @retriever = retriever
+    @url = url
+  end
+
+  def request(media_type)
+    etag_data, etag_response = etag_data_and_response(media_type) if first_page?
+    return log_skipped_requests(etag_data[:page_no]) if not_modified?(etag_response)
+
+    result = etag_response if etag_queried_for_first_page?(etag_data)
+    result ||= @retriever.do_request(@url, media_type)
+    store_etag_in_db(result) if cacheable_page?(result.meta['link'])
+    result
+  end
+
+  private
+
+  def etag_data_and_response(media_type)
+    etag_data = @retriever.db[:etags].first(base_url: base_url)
+    return unless etag_data
+    etag_response = get_etag_response(etag_data, media_type)
+    [etag_data, etag_response]
+  end
+
+  def not_modified?(response)
+    response && response.status[0] == '304'
+  end
+
+  def etag_queried_for_first_page?(etag_data)
+    etag_data && etag_data[:page_no] == 1
+  end
+
+  def first_page?
+    current_page_no.to_i == 1
+  end
+
+  def current_page_no
+    @current_page_no ||= extract_page_no(@url) || 1
+  end
+
+  def extract_page_no(string)
+    string.slice(/\bpage=(\d+)/, 1)
+  end
+
+  def store_etag_in_db(result)
+    record = @retriever.db[:etags].first(base_url: base_url)
```

Comment:
take a look at  insert_conflict http://sequel.jeremyevans.net/rdoc-adapters/classes/Sequel/Postgres/DatasetMethods.html#method-i-insert_conflict.  It will eliminate the need to query first and then insert.  Without an option it will just return if the record exists.

Reply:
Very nice

Reply:
However, it only works for postgres/sqlite. I will add it as a comment.

Reply:
Yes, I saw that.  I'm not sure if we should move forward with this until we are sure that Georgios will move to Postgres.  Unless we can check to see if the app is running mysql.